heron/spi/src/java/com/twitter/heron/spi/utils/ShellUtils.java

```diff
@@ -60,63 +60,80 @@ static String inputstreamToString(InputStream is) {
     return builder.toString();
   }
 
-  public static int runProcess(String[] cmdline, StringBuilder stdout, StringBuilder stderr) {
-    return runSyncProcess(false, false, cmdline, stdout, stderr, null);
+  public static int runProcess(String[] cmdline, StringBuilder stderr) {
+    return runSyncProcess(false, cmdline, stderr, null);
   }
 
   public static int runProcess(
-      boolean verbose, String cmdline, StringBuilder stdout, StringBuilder stderr) {
-    return runSyncProcess(verbose, false, splitTokens(cmdline), stdout, stderr, null);
+      String cmdline, StringBuilder stderr) {
+    return runSyncProcess(false, splitTokens(cmdline), stderr, null);
   }
 
   public static int runSyncProcess(
-      boolean verbose, boolean isInheritIO, String[] cmdline, StringBuilder stdout,
-      StringBuilder stderr, File workingDirectory) {
-    return runSyncProcess(isInheritIO, cmdline, stdout, stderr, workingDirectory,
+      boolean isInheritIO, String[] cmdline, StringBuilder stderr, File workingDirectory) {
+    return runSyncProcess(isInheritIO, cmdline, stderr, workingDirectory,
         new HashMap<String, String>());
   }
 
   /**
    * Start a daemon thread to read data from "input" to "out".
    */
-  private static Thread asyncProcessStream(final InputStream input, final StringBuilder out) {
+  private static Thread asyncProcessStream(final InputStream input, final StringBuilder builder) {
     Thread thread = new Thread() {
       @Override
       public void run() {
-        try {
-          out.append(inputstreamToString(input));
-        } finally {
+        // do not buffer
+        System.err.println("Process output (stdout+stderr):");
+        BufferedReader reader = new BufferedReader(new InputStreamReader(input), 1);
+        while(true) {
+          String line = null;
           try {
-            input.close();
+            line = reader.readLine();
           } catch (IOException e) {
-            LOG.log(Level.WARNING, "Failed to close the input stream", e);
+            LOG.log(Level.SEVERE, "Error when reading line from subprocess", e);
+          }
+          if (line == null) {
+            break;
+          } else {
+            System.err.println(line);
+            builder.append(line);
           }
         }
+        try {
+          input.close();
+        } catch (IOException e) {
+          LOG.log(Level.WARNING, "Failed to close the input stream", e);
+        }
       }
     };
     thread.setDaemon(true);
-    thread.start();
```

Comment:
please clarify in the javadocs that this method doesn't actually run anything. We should rename the method to make that very obvious too actually. Maybe createAsyncProcessThread?